File: Torch/MaxPooling.lua
-- Copyright 2016-present, Facebook, Inc.
-- All rights reserved.
--
-- This source code is licensed under the license found in the
-- LICENSE file in the root directory of this source tree.

return function(sparseconvnet)
  local C = sparseconvnet.C
  local MaxPooling, parent = torch.class(
    'sparseconvnet.MaxPooling', 'nn.Module', sparseconvnet)

  function MaxPooling:__init(
      dimension, poolSize, poolStride, nFeaturesToDrop)
    parent.__init(self)
    self.dimension = dimension
    self.poolSize = sparseconvnet.toLongTensor(poolSize,dimension)
    self.poolStride = sparseconvnet.toLongTensor(poolStride,dimension)
    self.poolVolume = self.poolSize:prod()
    self.nFeaturesToDrop = nFeaturesToDrop or 0
    self.output = {
      features = torch.FloatTensor(),
    }
    self.gradInput = {
      features = torch.Tensor()
    }
  end

  function MaxPooling:updateOutput(input)
    self.output.metadata=input.metadata
    self.output.spatialSize
    = torch.cdiv(input.spatialSize-self.poolSize,self.poolStride)+1
    C.dimTypedFn(self.dimension, self._type, 'MaxPooling_updateOutput')(
      input.spatialSize:cdata(),
      self.output.spatialSize:cdata(),
      self.poolSize:cdata(),
      self.poolStride:cdata(),
      input.metadata.ffi,
      input.features:cdata(),
      self.output.features:cdata(),
      self.nFeaturesToDrop,
      self.shared.rulesBuffer and self.shared.rulesBuffer:cdata())
    return self.output
  end

  function MaxPooling:updateGradInput(input, gradOutput)
    C.dimTypedFn(self.dimension, self._type, 'MaxPooling_updateGradInput')(
      input.spatialSize:cdata(),
      self.output.spatialSize:cdata(),
      self.poolSize:cdata(),
      self.poolStride:cdata(),
      input.metadata.ffi,
      input.features:cdata(),
      self.gradInput.features:cdata(),
      self.output.features:cdata(),
      gradOutput.features:cdata(),
      self.nFeaturesToDrop,
      self.shared.rulesBuffer and self.shared.rulesBuffer:cdata())
    return self.gradInput
  end

  function MaxPooling:type(type,tensorCache)
    if type==nil then
      return self._type
    end
    self._type=type
    self.output.features=self.output.features:type(type)
    self.gradInput.features=self.gradInput.features:type(type)
    --return parent.type(self,type,tensorCache)
  end

  function MaxPooling:__tostring()
    local s = 'MaxPooling'
    if self.poolSize:max()==self.poolSize:min()
    and self.poolStride:max()==self.poolStride:min() then
      s=s..self.poolSize[1] ..(self.poolStride[1]==1
        and '' or '/'..self.poolStride[1])
    else
      s=s..'('..self.poolSize[1]
      for i=2,self.dimension do
        s=s..','..self.poolSize[i]
      end
      s=s..')/('..self.poolStride[1]
      for i=2,self.dimension do
        s=s..','..self.poolStride[i]
      end
      s=s..')'
    end
    if self.nFeaturesToDrop>0 then
      s=s .. ' nFeaturesToDrop = ' .. self.nFeaturesToDrop
    end
    return s
  end

  function MaxPooling:clearState()
    self.output = {
      features = self.output.features:set(),
    }
    self.gradInput = {
      features = self.gradInput.features:set()
    }
  end

  function MaxPooling:suggestInputSize(nOut)
    return torch.cmul(nOut-1,self.poolStride)+self.poolSize
  end
end

File: PyTorch/sparseconvnet/legacy/maxPooling.py
# Copyright 2016-present, Facebook, Inc.
# All rights reserved.
#
# This source code is licensed under the license found in the
# LICENSE file in the root directory of this source tree.

import torch
import sparseconvnet
from . import SparseModule
from ..utils import toLongTensor, dim_typed_fn, optionalTensor, nullptr
from ..sparseConvNetTensor import SparseConvNetTensor


class MaxPooling(SparseModule):
    def __init__(self, dimension, pool_size, pool_stride, nFeaturesToDrop=0):
        SparseModule.__init__(self)
        self.dimension = dimension
        self.pool_size = toLongTensor(dimension, pool_size)
        self.pool_stride = toLongTensor(dimension, pool_stride)
        self.pool_volume = self.pool_size.prod()
        self.nFeaturesToDrop = nFeaturesToDrop or 0
        self.output = SparseConvNetTensor(torch.Tensor())
        self.gradInput = torch.Tensor()

    def updateOutput(self, input):
        self.output.metadata = input.metadata
        self.output.spatial_size =\
            (input.spatial_size - self.pool_size) / self.pool_stride + 1
        dim_typed_fn(self.dimension, input.features, 'MaxPooling_updateOutput')(
            input.spatial_size,
            self.output.spatial_size,
            self.pool_size,
            self.pool_stride,
            input.metadata.ffi,
            input.features,
            self.output.features,
            self.nFeaturesToDrop,
            torch.cuda.IntTensor() if input.features.is_cuda else nullptr)
        return self.output

    def updateGradInput(self, input, gradOutput):
        dim_typed_fn(self.dimension, input.features, 'MaxPooling_updateGradInput')(
            input.spatial_size,
            self.output.spatial_size,
            self.pool_size,
            self.pool_stride,
            input.metadata.ffi,
            input.features,
            self.gradInput,
            self.output.features,
            gradOutput,
            self.nFeaturesToDrop,
            torch.cuda.IntTensor() if input.features.is_cuda else nullptr)
        return self.gradInput

    def type(self, t=None, tensorCache=None):
        if t is None:
            return self._type
        self.output.type(t)
        self.gradInput = self.gradInput.type(t)

    def __repr__(self):
        s = 'MaxPooling'
        if self.pool_size.max() == self.pool_size.min() and\
                self.pool_stride.max() == self.pool_stride.min():
            s = s + str(self.pool_size[0]) + '/' + str(self.pool_stride[0])
        else:
            s = s + '(' + str(self.pool_size[0])
            for i in self.pool_size[1:]:
                s = s + ',' + str(i)
            s = s + ')/(' + str(self.pool_stride[0])
            for i in self.pool_stride[1:]:
                s = s + ',' + str(i)
            s = s + ')'

        if self.nFeaturesToDrop > 0:
            s = s + ' nFeaturesToDrop = ' + self.nFeaturesToDrop
        return s

    def suggestInputSize(self, out_size):
        return (out_size - 1) * self.pool_stride + self.pool_size
